pennylane/devices/qubit/apply_operation.py

```diff
@@ -263,6 +263,43 @@ def apply_cnot(op: qml.CNOT, state, is_state_batched: bool = False, debugger=Non
     return math.stack([state[sl_0], state_x], axis=control_axes)
 
 
+@apply_operation.register
+def apply_grover(op: qml.GroverOperator, state, is_state_batched: bool = False, debugger=None):
+    r"""Apply GroverOperator to state. This method uses that this operator
+    is :math:`2*P-\mathbb{I}`, where :math:`P` is the projector onto the
+    all-plus state. This allows us to compute the new state by replacing summing
+    over all axes on which the operation acts, and "filling in" the all-plus state
+    in the resulting lower-dimensional state via a Kronecker product.
+    """
+    n_wires = len(op.wires)
+    if n_wires < 9:
+        return _apply_operation_default(op, state, is_state_batched, debugger)
+
+    # 2 * Squared normalization of the all-plus state on the op wires
+    # (squared, because we skipped the normalization when summing, 2* because of the def of Grover)
+    prefactor = 2 ** (1 - n_wires)
+
+    # The axes to sum over in order to obtain <+|\psi>, where <+| only acts on the op wires.
+    sum_axes = [w + is_state_batched for w in op.wires]
+    collapsed = math.sum(state, axis=tuple(sum_axes))
+
+    if n_wires == (len(qml.math.shape(state)) - is_state_batched):
+        # If the operation acts on all wires, we can skip the tensor product with all-ones state
+        new_shape = (-1,) + (1,) * n_wires if is_state_batched else (1,) * n_wires
+        return prefactor * math.reshape(collapsed, new_shape) - state
+        # [todo]: Once Tensorflow support expand_dims with multiple axes in the second argument,
+        # use the following line instead of the two above.
+        # return prefactor * math.expand_dims(collapsed, sum_axes) - state
+
+    all_plus = math.cast_like(math.full([2] * n_wires, prefactor), state)
+    # After the Kronecker product (realized with tensordot with axes=0), we need to move
+    # the new axes to the summed-away axes' positions. Finally, subtract the original state.
+    source = list(range(math.ndim(collapsed), math.ndim(state)))
+    # Probably it will be better to use math.full or math.tile to create the outer product
+    # here computed with math.tensordot. However, Tensorflow and Torch do not have full support
+    return math.moveaxis(math.tensordot(collapsed, all_plus, axes=0), source, sum_axes) - state
```

Comment:
If we pulled this out into an `_apply_specialized_grover` function, we would then be able to validate the general process on fewer wires, which would help us a have confidence in at larger system sizes.

Reply:
agreed, would be nice. we can follow the code standard set by `apply_parametrized_evolution` to be consistent 👍

Reply:
Done :)